pdns/syncres.hh

```diff
@@ -78,60 +78,53 @@ typedef std::unordered_map<
 template<class Thing> class Throttle : public boost::noncopyable
 {
 public:
-  Throttle() : d_limit(3), d_ttl(60), d_last_clean(time(nullptr))
-  {
-  }
 
-  struct entry
+  struct entry_t
   {
+    Thing thing;
     time_t ttd;
-    unsigned int count;
+    mutable unsigned int count;
   };
-  typedef map<Thing,entry> cont_t;
+  typedef multi_index_container<entry_t,
+                                indexed_by<
+                                  ordered_unique<tag<Thing>, member<entry_t, Thing, &entry_t::thing>>,
+                                  ordered_non_unique<tag<time_t>, member<entry_t, time_t, &entry_t::ttd>>
+                                  >> cont_t;
 
-  bool shouldThrottle(time_t now, const Thing& t)
+  bool shouldThrottle(time_t now, const Thing &t)
   {
-    if(now > d_last_clean + 300 ) {
-
-      d_last_clean=now;
-      for(typename cont_t::iterator i=d_cont.begin();i!=d_cont.end();) {
-        if( i->second.ttd < now) {
-          d_cont.erase(i++);
-        }
-        else
-          ++i;
-      }
-    }
-
-    typename cont_t::iterator i=d_cont.find(t);
-    if(i==d_cont.end())
+    auto i = d_cont.find(t);
+    if (i == d_cont.end()) {
       return false;
-    if(now > i->second.ttd || i->second.count == 0) {
+    }
+    if (now > i->ttd || i->count == 0) {
       d_cont.erase(i);
       return false;
     }
-    i->second.count--;
+    i->count--;
 
     return true; // still listed, still blocked
   }
-  void throttle(time_t now, const Thing& t, time_t ttl=0, unsigned int tries=0)
-  {
-    typename cont_t::iterator i=d_cont.find(t);
-    entry e={ now+(ttl ? ttl : d_ttl), tries ? tries : d_limit};
 
-    if(i==d_cont.end()) {
-      d_cont[t]=e;
+  void throttle(time_t now, const Thing &t, time_t ttl, unsigned int count)
+  {
+    auto i = d_cont.find(t);
+    time_t ttd = now + ttl;
+    if (i == d_cont.end()) {
+      entry_t e = { t, ttd, count };
+      d_cont.insert(e);
+    } else if (i->ttd > ttd || i->count < count) { // ????
+      auto &ind = d_cont.template get<Thing>();
+      ind.modify(i, [ttd,count](entry_t &e) { e.ttd = ttd; e.count = count; });
     }
-    else if(i->second.ttd > e.ttd || (i->second.count) < e.count)
-      d_cont[t]=e;
```

Comment:
@rgacogne can you explain the conditions used to decide to modify or not?

Reply:
IMHO we should always replace by the maximum of the TTDs and the maximum of the number of attempts. It looks like the existing code replaces if the new TTD is smaller or equal to the existing one which doesn't make sense to me.

Reply:
Good, I also could not make any sense of the existing condition.